Add toggle to fix module_defaults with module-as-redirected-action on a per-module basis

s-hertel opened the pull request:
##### SUMMARY

Possible solution for #77059.

This adds a new `action_plugin` routing field for modules to support modules that need to use a common action plugin but don't want module_defaults resolved to the action plugin.

Instead of hijacking the action plugin redirection to ensure a platform-specific handler is used for a module:

```
plugin_routing:
  action:
    somemodule:  # Was never an action plugin, just here to replicate platform_somemodule prefix magic for determining the action plugin
      redirect: testns.testcoll.platform
  modules:
    somemodule:
      redirect: testns.testcoll.platform_somemodule
```

The resolved module can specify an action plugin directly:

```
plugin_routing:
  modules:
    somemodule:
      redirect: testns.testcoll.platform_somemodule
    platform_somemodule:
      action_plugin: testns.testcoll.platform
```

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request
- Feature Pull Request
- Test Pull Request

##### COMPONENT NAME
lib/ansible/executor/task_executor.py
lib/ansible/playbook/base.py

s-hertel committed:
If there is a platform specific handler, prefer the resolved module over the resolved action when loading module_defaults

Add a toggle for action plugins to prefer the resolved module when loading module_defaults

Allow moving away from modules intercepted as actions pattern

Fixes #77059

s-hertel committed:
update unit tests

s-hertel committed:
add integration tests

s-hertel committed:
fix integration tests

s-hertel committed:
rename method

s-hertel committed:
Change toggle to module routing field 'use_action' and update tests

Add backwards-compat test for module-as-redirected-actions getting action-specific defaults

bcoca commented on test/integration/targets/module_defaults/collections/ansible_collections/testns/testcoll/meta/runtime.yml:
```diff
@@ -1,3 +1,38 @@
+
+    # Not "eos_facts" to ensure TE is not finding handler via prefix
+    # eosfacts tasks should not get eos module_defaults (or defaults for other modules that use eos action plugin)
+    eosfacts:
```
I would add another test with eos_facts/ios_facts to see the 'other matching path' working

bcoca replied:
can also be added to test_defaults with collections: , so we get task with the non fqcn name

s-hertel replied:
I added better tests for ios_facts with the collections keyword. Now the test suite verifies that calling the task with either iosfacts or ios_facts will get module_defaults entries `testns.testcoll.iosfacts`/`testns.testcoll.ios_facts`.

I also removed the changes to make more stuff rely on the prefix magic. Now this is fixed only via toggle, so the runtime.yml changed a bit to match the tests' expectations.

s-hertel committed:
Remove new prefix magic

Update tests accordingly

Add tests with the collections keyword

test module action_plugin redirects

s-hertel committed:
changelog

s-hertel committed:
fix diff and add docstring

s-hertel committed:
Use a more realistic example in changelog

nitzmahone commented on lib/ansible/executor/task_executor.py:
```diff
@@ -1106,8 +1116,16 @@ def _get_action_handler(self, connection, templar):
 
         collections = self._task.collections
 
+        # Check if the module has specified to action handler
```
s/to/an/

nitzmahone commented on lib/ansible/executor/task_executor.py:
```diff
@@ -1106,8 +1116,16 @@ def _get_action_handler(self, connection, templar):
 
+        # Check if the module has specified to action handler
+        module = None
+        if self._shared_loader_obj.module_loader.has_plugin(self._task.action, collection_list=collections):
```
So this is technically a breaking change, since we're now favoring modules over actions, at least when they declare an overridden action plugin (where previously, we always looked for actions first). I think *in practice* it's probably fine, and I assumed we'd probably have to do it this way at some point anyway. This still preserves the common-ish case of an action + module pair with the same names favoring the action, so long as the module doesn't declare an action_plugin in its runtime metadata (in which case the action would've previously masked it anyway). 

Is the only reason for switching the order to (sort of) accommodate recent collections that were using action redirects for this purpose? If so, I suspect those collections would require the new keyword to work anyway, so maybe it's not worth the risk. I know I'd mentioned swapping the order at one point, but that was only to try and make it work without the new keyword (which would've failed in some cases anyway). Since we have the new keyword, can't we just basically use that with the previous action->module resolution order and just explicitly override the action handler when the new keyword is present? That would make the change a lot less risky and much more palatable to backport...

Also, if we keep it this way, this case should probably be collapsed to a single hit to the plugin loader, since `has_plugin` is internally already calling `find_plugin_with_context`.

s-hertel replied:
This change is entirely to support the new module routing keyword. Since `_load_module_defaults` now does not merge all module defaults to the same action plugin name, TE needs to access the defaults using the same name the defaults were stored under and ensure the action_plugin specified by the module gets used. It shouldn't be a breaking change unless someone decided to add a action_plugin entry for module routing before it did anything. I could tighten this up a little to set the module context back to None if it isn't resolved or defined the action plugin it needs:
```diff
diff --git a/lib/ansible/executor/task_executor.py b/lib/ansible/executor/task_executor.py
index 6e699285b0..2c51f8ee32 100644
--- a/lib/ansible/executor/task_executor.py
+++ b/lib/ansible/executor/task_executor.py
@@ -1122,7 +1122,9 @@ class TaskExecutor:
             module = self._shared_loader_obj.module_loader.find_plugin_with_context(
                 self._task.action, collection_list=self._task.collections
             )
-        if module is not None and module.resolved and module.action_plugin:
+        if module is None or not module.resolved or not module.action_plugin:
+            module = None
+        if module is not None:
```
I'll try to consolidate this to just one call as well - I can't remember why I did this, perhaps there's an error if the module does not exist (in which case it falls back to existing behavior).

s-hertel replied:
It does seem like an odd change to assume the module name here/_load_module_defaults when the action plugin may not respect it. Initially I had a `prefer_module_args` field for action plugin routing (instead of `action_plugin` for module routing), but it was still necessary for _load_module_defaults/TE to make a guess about the module name the action plugin would use.

bcoca replied:
it really does not break backwards compat unless people were using the new key already

nitzmahone reviewed (changes requested):
Change and tests look good, though we should talk through the resolution order change in the playbook resolver and task executor to make sure that's truly necessary before committing to it. It probably doesn't have a huge practical effect, but it is a breaking change that would be a lot more risky to backport IMO.

s-hertel committed:
Set module back to None if the module routing hasn't specified an action plugin

Remove extra has_plugin call and update unit tests

fix typo

s-hertel commented on test/units/executor/test_task_executor.py:
```diff
@@ -204,8 +208,11 @@ def test_task_executor_get_action_handler(self):
+        te._shared_loader_obj.module_loader.find_plugin_with_context.return_value = context
         action_loader = te._shared_loader_obj.action_loader
         action_loader.has_plugin.return_value = True
+
```
Fixing diff
```suggestion

```

s-hertel committed:
Update test/units/executor/test_task_executor.py

nitzmahone reviewed (approved):
OK, I tried every combo I could think of that would end up with a user-visible change that wasn't already broken and couldn't come up with anything else, so +1 from me to merge. Thanks for continuing to beat on this one!

s-hertel commented:
Thanks for the reviews! :tada:
